---
title: "Prediction/Discovery Tutorial from GDC TCGA Data"
output:
  html_notebook:
    df_print: paged
editor_options:
  chunk_output_type: inline
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# Set up package environment

We first want to install all of the packages used in this analysis. We use the package `renv` as a way to manage package versions. The following code was initially run to set up the package library.

```{r renv, eval = FALSE}
install.packages("renv")
library(renv)

renv::init(bare = TRUE)
install.packages("BiocManager")
library(BiocManager)
install.packages(c("BSgenome", "rmarkdown", "knitr", "TCGAbiolinks", 
                   "devtools"), repo = BiocManager::repositories())
devtools::install_github("campbio/musicatk@v1.0.1")
renv::snapshot()
```

If you are attempting to reproduce this analysis, you can 1) install/load `renv`, 2) copy/paste the `renv.lock` file into your current working directory and 3) run the `renv::restore()` command to automatically install all of the packages with the same version. 

```{r example_restore, eval = FALSE}
# Make sure the 'renv.lock' file is in your current working directory.
install.packages("renv")
library(renv)
renv::restore()
```

# Load required libraries

```{r library, message=FALSE}
library("musicatk")
library("TCGAbiolinks")
library("withr")
```

# Import TCGA datasetes using TCGAbiolinks
Mutation calls in maf format will be downloaded for sequencing datasets from The Cancer Genome Atlas (TCGA) consortium. These variants were generated by the NCI's Genomic Data Commons (GDC) and can be retrieved using the package `TCGAbiolinks`. For this tutorial we use the four smallest datasets from TCGA, in order to show an analysis that can be performed without much waiting for compute.

```{r import, message=FALSE, results=FALSE}
tcga_datasets <- c("TCGA-CHOL", "TCGA-DLBC", "TCGA-UCS", "TCGA-KICH")

types <- sapply(strsplit(tcga_datasets, "-"), '[', 2)
dataset <- NULL
annot <- NULL
for(i in seq_along(tcga_datasets)) {
  query <- GDCquery(project = tcga_datasets[i], 
                  data.category = "Simple Nucleotide Variation", 
                  data.type = "Masked Somatic Mutation", 
                  workflow.type = "MuTect2 Variant Aggregation and Masking",
                  experimental.strategy = "WXS",
                  data.format = "maf")
  GDCdownload(query)
  data <- GDCprepare(query)
  variants <- extract_variants_from_matrix(data)
  dataset <- rbind(dataset, variants)  
  annot <- rbind(annot, cbind(rep(types[i], length(unique(variants$sample))), 
                              unique(as.character(variants$sample))))
}
```


# Create musica object

We need to look up context bases in order to build tables, and do to that we need a representation of the genome we're working with. TCGA from GDC was aligned with HG38, so we can create an HG38 BSgenome object using select_genome. Then we create a musica object using create_musica with the dataset from GDC and the genome object we selected.

create_musica performs some basic checks and preprocessing to ensure that the genomes match, insertions and deletions are in a uniform style, and converts adjacent single base substitutions into double base substitutions, also for standard processing. All of these options can be enabled or disabled using flags within the function.

```{r create_musica}
g <- select_genome("38")
tcga <- create_musica(dataset, g)
```

# Add sample annotation

We match tumor types to sample names in order to perform advanced plotting and downstream analysis later in the workflow. Here we're just matching sample names in the variants matrix to the sample names in the annotation matrix, and then add them to the musica object using the samp_annot command and name the annotation "Tumor_Types".

```{r add_annotation}
matched_annot <- annot[match(samp_annot(tcga)$Samples, annot[,2]), 1]
samp_annot(tcga, "Tumor_Types") <- matched_annot
```

# SBS Umap

This should be split up into different sections (build table & subset, prediction, plotting)

We build the standard single base substitution (SBS) table with one context base on either side, 96 total motifs. We simply select "SBS96" as our choice of table to build.
```{r build_tables, message=FALSE}
build_standard_table(tcga, g, "SBS96")
```

This step is optional, but it can improve signal overall. We subset our cohor to only samples with at least 5 SBS96 counts.
```{r subset}
tcga_sbs_subset <- subset_musica_by_counts(musica = tcga, table_name = "SBS96", num_counts = 5)
```

We use existing signatures to predict exposures for our set of samples. We're using the latest COSMIC v3 results, we're matching to the SBS data, because we are using single base substitutions (SBS), and because our samples are whole exome sequencing (WES) rather than whole genome sequencing (WGS), we use the exome signatures. 

The auto_predict grid does two things for us automatically. Firstly, it predicts signatures for each tumor type separately, as they have very different profiles, and then combines them together at the end. Secondly, it subsets to only the signatures that are active in the dataset using a 2-stage heuristic process. This makes it easier to sort through and characterize the relevant signature to our specific cohort. We use latent dirichlet (lda) prosterior for the prediction.
```{r predict}
tcga_v3_sbs <- with_seed(1, auto_predict_grid(musica = tcga_sbs_subset, 
                                              table_name = "SBS96", 
                                              signature_res = cosmic_v3_sbs_sigs_exome, 
                                              algorithm = "lda",
                                              sample_annotation = "Tumor_Types"))
```

We can now start to analyze our prediction results. We can use plot_exposures to view the exposure of each sample to each signature. We appear to have 14 signatures, and a wide range of raw counts for our samples, with one sample having over 5000 counts. In order to view things a little bit more clearly, we can add the proportional parameter and set it to TRUE.

At this point we can choose a few signatures to view more closely. We can sort by a single or a set of signatures to effectively view patterns in the exposures. Let's start with SBS46. We use the sort_samples parameter with the selected signature name. We can sort by multiple signatures, so let's add SBS9 and SBS13 as well.

We can make use of advanced plotting with our annotation we added earlier. Let's group by annotation, and select the "Tumor_Types" annotation. We can see that SBS9 is found in DLBC, SBS46 is found in KICH, and SBS13 is found in UCS.
```{r plot_prediction_exposures}
plot_exposures(tcga_v3_sbs, proportional = TRUE, sort_samples = c("SBS46", "SBS9", "SBS13"), group_by = "annotation", annotation = "Tumor_Types")
```
Uniform Maniform Approximation and Projection (UMAP) is a technique for visualization and dimension reduction. We use it here to view our samples exposures mapped on a 2-d grid. We color them by the "Tumor_Types" annotation. We can see a good separation between the tumor types.
```{r plot_umap, echo=FALSE, fig.align="center", fig.height=2, fig.width=2}
with_seed(1, create_umap(tcga_v3_sbs))
plot_umap(result = tcga_v3_sbs, proportional = TRUE, color_by = "annotation", annotation = "Tumor_Types", add_annotation_labels = TRUE, annotation_text_box = TRUE, annotation_label_size = 6, legend = FALSE, strip_axes = TRUE)
```

We can also plot the same umap, this time colored by exposure to every different signature. This allows us to associate signatures with specific tumor types. For instance SBS19 and SBS25 seem to be found only in DLBC. SBS32 is only found in KICH. And SBS5 is found everywhere.
```{r plot_umap_sigs, fig.height=6, fig.width=10, fig.align="center"}
plot_umap(tcga_v3_sbs, same_scale = FALSE)
```


# Discover Signatures
Instead of predicting exposures from existing signature, we're now discovering exposures and signatures simultaneously. We use the same tcga_sbs_subset we used earlier, with the same SBS96 table, but now for de-novo discover. We need to a select a number of signatures (k), which is actually a complicated process with too low a k leading to missed signature, and too low a k leading to overfitting. In this case we'll choose k=10 because it's a bit lower than the number of signatures that were determined to be active using prediction.
```{r discovery}
denovo <- discover_signatures(tcga_sbs_subset, "SBS96", num_signatures=10, nstart=1)
```

We can plot the new signatures that we've discovered. It's promising that the signatures look distinct from each other, they're not all flat distributions. 
```{r plot_signatures}
plot_signatures(denovo)
```
We can compare these new signatures to existing signatures from COSMIC. We'll use COSMIC v2 
as an example, because there are fewer signatures total to compare to, making it a bit simpler. 
We can see that we've matched with three COSMIC signatures using the default threshold of 0.9 cosine similarity metric. On the left are our novel signatures, 5, 7, 9; on the right are the matching COSMIC signatures 10, 1, and 6. These signatures have known etiologies corresponding to altered activity of the error-prone polymerase POLE, spontaneous deamination of 5-methylcytosine, and defective DNA mismatch repair, respectively. We can look at the table as well and see the exact values for signatures matching.
```{r plot_comparison}
compare_cosmic_v2(denovo)
```
To view trends across samples and across tumor types for multiple signatures, we can plot the exposures to our discovered signatures. We set proportional to scale samples onto one axis, we group by annotation and set that annotation to "Tumor_Types" to split our samples out by tumor type. We can sort by a given signature (we'll choose "Signature2") to see patterns emerge in our sample exposures. For instance, KICH has a very flat exposure spectrum, but UCS has a few samples with extremely high exposure to this signature.
```{r plot_exposures}
plot_exposures(denovo, proportional = TRUE, sort_samples = "Signature2", group_by = "annotation", annotation = "Tumor_Types")
```
Musicatk can create heatmap plots that hierarchically cluster samples based on signature exposure. We can additionally label samples by their tumor type, allowing us to see associations between tumor types and samples. For instance, UCS appears to be correlated with Signature7, and DLBC appears to be correlated with Signature4. 
```{r plot_heatmap}
plot_heatmap(res_annot = denovo, proportional = TRUE, scale = TRUE, annotation = "Tumor_Types")
```
To see this comparison directly we can perform a differential analysis. Plotting the results we can see that indeed USC is linearly related to Signature7, although not as strongly as signature 9. And DLBC is related Signature 4 and also to Signature7.
```{r diff_analysis, warning=FALSE}
glm_stats <- exposure_differential_analysis(musica_result = denovo, 
                                            annotation = "Tumor_Types", 
                                            method = "glm.nb")
plot_differential_analysis(glm_stats, "glm", 4)
```
