---
title: "Association of mutation rates with WGD status"
author: "Joshua D. Campbell"
date: "`r format(Sys.time(), '%B %d, %Y')`"
output:
  html_document:
    number_sections: yes
    toc: yes
  html_notebook: default
---

# Set up package environment

We first want to install all of the packages used in this analysis. We use the package `renv` as a way to manage package versions. The following code was initially run to set up the package library.

```{r renv, eval = FALSE}
install.packages("renv")
library(renv)

renv::init(bare = TRUE)
install.packages("BiocManager")
library(BiocManager)
install.packages(c("ggplot2", "magrittr", "knitr", "kableExtra", "ggrepel",
                   "rmarkdown", "formatR", "bitops", "caTools","cowplot",
                   "Rcpp","digest","evaluate","htmltools","jsonlite",
                   "base64enc","rprojroot","highr","xfun","stringi"),
                 repo = BiocManager::repositories())
renv::snapshot()
```

If you are attempting to reproduce this analysis, you can 1) install/load `renv`, 2) copy/paste the `renv.lock` file into your current working directory and 3) run the `renv::restore()` command to automatically install all of the packages with the same version. 

```{r example_restore, eval = FALSE}
# Make sure the 'renv.lock' file is in your current working directory.
install.packages("renv")
library(renv)
renv::restore()
```

# Load in scripts and R packages

Next, R packages will be loaded and scripts will be sourced.

```{r library, message = FALSE}
source("logistic_regression_util.R")
source("maf_utils.R")
library("ggplot2")
library("cowplot")
library("magrittr")
library("knitr")
library("kableExtra")
library("ggrepel")
knitr::opts_chunk$set(fig.width=12, fig.height=8)
```



# Read in data files

The file containing Mutation burden (i.e. mutation per Mb) and MSI or POLE status was obtained from the DNA damage TCGA working group from the PanCanAtlas consortium. This work was published in Cell and can be found at:

`https://www.cell.com/cell-reports/pdf/S2211-1247(18)30437-6.pdf`.

Data from the paper can be found at:

`https://gdc.cancer.gov/about-data/publications/PanCan-DDR-2018`.

The ABSOLUTE call file can be found at:

`http://api.gdc.cancer.gov/data/4f277128-f793-4354-a13d-30cc7fe9f6b5`. 

## Read in tumor annotation
```{r read_files}
# Read in mutation rate data
mut.rates = read.table("../../Data/Mutation_Rate_and-MSI.txt", header=TRUE, stringsAsFactors = FALSE, sep="\t", row.names = 1)

# Read in ABSOLUTE calls and combine with mut.rates
absolute = read.table("../../Data/TCGA_mastercalls.abs_tables_JSedit.fixed.txt", header = TRUE, stringsAsFactors = FALSE, sep = "\t", row.names = 1)
mut.rates$Ploidy = absolute[rownames(mut.rates), "ploidy"]
mut.rates$PloidyCorrected_MutationPerMb = log10(mut.rates$Non.silentMutationsperMb) / mut.rates$Ploidy

mut.rates$WGD = factor(ifelse(mut.rates$Genome_doublings > 0, "WGD", "Non-WGD"), levels = c("Non-WGD", "WGD"))
mut.rates$MSI_POLE_Status = factor(ifelse(mut.rates$MSIPOLE == "MSIPOLE", "Yes", "No"), levels = c("No", "Yes"))
mut.rates$Log10_MutsPerMB = log10(mut.rates$Non.silentMutationsperMb)

# Make tumor type a factor ordered by median mutation rate
agg = aggregate(mut.rates$Non.silentMutationsperMb,
                by=list(mut.rates$Type),
                "median")
tumor.type = agg[order(agg[,2]),1]
mut.rates$Type = factor(mut.rates$Type, levels = tumor.type)

# Create subsets
mut.rates.hyper = subset(mut.rates, MSI_POLE_Status == "Yes")
mut.rates.nonhyper = subset(mut.rates, MSI_POLE_Status == "No")
```

## Read in and process gene mutations

All tumor ids in `mut.rates` match an entry in the PanCanAtlas MC3 maf, however some of the tumors are no longer present after filtering out poor quality variants. Therefore, the `Tumor_Sample_Barcode` column within the maf is converted to a factor and the levels are set to the tumors found in `mut.rates` so each tumor will get a column in the `muts` table generated by `xtabs`. The maf file was obtained from `http://api.gdc.cancer.gov/data/1c8cfe5f-e52d-41ba-94da-f15ea1337efc`. PanCan MutSig2CV results were obtained from `https://api.gdc.cancer.gov/data/2e8d4650-5498-4288-a461-23667ff553e2`. 

```{r process_maf}
# Read in file and subset to filtered variants
maf = read.table(gzfile("../../Data/mc3.v0.2.8.PUBLIC.maf.gz"), sep="\t", header=TRUE, stringsAsFactors = FALSE, quote="")
maf = subset(maf, FILTER %in% c("PASS", "wga", "native_wga_mix"))

# Substring ID and convert to factor with appropriate levels matching all tumors in mut.rates
maf$Tumor_Sample_Barcode = factor(substring(maf$Tumor_Sample_Barcode, 1, 15), levels = rownames(mut.rates))

# Subset to nonsynonymous variants
maf.c = maf.coding(maf)
muts = xtabs(~ Hugo_Symbol + Tumor_Sample_Barcode, data = maf.c)

# Subset further to significantly mutated genes from MutSig2CV
fs = list.files(path="../../Data/MutSig2CV/", full.names = TRUE)
fs = setdiff(fs, "../../Data/MutSig2CV//README.txt")
genes = c()
for(i in fs) {
  temp = read.table(i, header=TRUE, stringsAsFactors=FALSE, sep="\t", quote="")
  temp = subset(temp, qvalue < 0.1)
  genes = c(genes, temp[,1])
}
genes = sort(intersect(unique(genes), rownames(muts)))

muts = muts[genes,]

# Clear up some memory
rm(maf)
rm(maf.c)
```



# Plotting distribution of mutation rates

We need to examine the distribution of mutation rates to determine appropriate statistical tests. In parts A and B, the density of non-silent mutations is plotted across `r nrow(mut.rates)` tumors from TCGA. Densities are plotted separately for tumors with MSI or POLE tumors, which are expected to have higher mutation rates in general. By applying a log10 transformation in part B, we can more clearly see the distubtions. 

In parts B and C, the density of non-silent mutations is plotted across `r nrow(mut.rates.nonhyper)` tumors from TCGA that do <b>NOT</b> fall into MSI or POLE categories. In part C, we can see that although hypermutated samples are excluded, there is still a long tailed distribution. The application of log10 in part D makes the distrubution appear more "bell-shapped". Therefore, we should either use non-parametric tests (e.g. Wilcoxon) or use logged transformed values with parametric tests (e.g. t-test).

```{r ploting_distributions}
p1 = ggplot(mut.rates, aes(x=Non.silentMutationsperMb, fill=MSI_POLE_Status)) + geom_density(alpha = 0.5) + theme_bw() + ggtitle("All tumors, no log transformation")
p2 = ggplot(mut.rates, aes(x=Non.silentMutationsperMb, fill=MSI_POLE_Status)) + geom_density(alpha = 0.5) + theme_bw() + ggtitle("All tumors, with Log10 transformation") + scale_x_log10()

p3 = ggplot(mut.rates.nonhyper, aes(x=Non.silentMutationsperMb, fill=MSI_POLE_Status)) + geom_density(alpha = 0.5) + theme_bw() + ggtitle("Non-hypermutated tumors, no log transformation")
p4 = ggplot(mut.rates.nonhyper, aes(x=Non.silentMutationsperMb, fill=MSI_POLE_Status)) + geom_density(alpha = 0.5) + theme_bw() + ggtitle("Non-hypermutated tumors, with Log10 transformation") + scale_x_log10()

plot_grid(p1, p2, p3, p4, labels = "AUTO")
```



# Testing associations between mutation burden and WGD status

Next, we will use linear modeling to examine associations between mutation burden per Mb and WGD status while controlling for tumor type and MSI/POLE status. Note that in the "Estimate" column in the "Coefficients" section within the summary tables, the `(Intercept)` is the baseline group and refers to the mean mutation rate for `r print(tumor.type[1])`, non-WGD, non-MSI/POLE tumors. `WGDWGD` refers to the difference between the baseline group and WGD tumors. `MSI_POLE_StatusYes` refers to the difference between the baseline group and tumors with MSI/POLE status. Rows for each tumor type (e.g. `TypeBLCA`) refer to the mean difference between each tumor type and the baseline group.

There are no significant associations between WGD status and mutation rate when using the non-transformed values. However, when using log-transformed values, we see strong associations with WGD status. This is true when including MSI/POLE tumors and controlling for this in the model AND when subsetting to non-MSI/POLE tumors and estimating a new model.

Overall, these results show that there is an overall strong association between WGD status and mutation rate <b> when averaging across all tumor types </b>. This analysis does not show whether this trend is true for all tumor types or if a subset of tumor types are driving the association.

```{r mut_rates}
# Controlling for tumor type and Hypermutation
model.full = lm(Non.silentMutationsperMb ~ WGD + MSI_POLE_Status + Type, data=mut.rates)
summary(model.full)

# Controlling for tumor type and Hypermutation with log transformation
model.full.log10 = lm(log10(Non.silentMutationsperMb) ~ WGD + MSI_POLE_Status + Type, data=mut.rates)
summary(model.full.log10)

# Testing hypermutated samples only
model.hyper = lm(Non.silentMutationsperMb ~ WGD + Type, data=mut.rates.hyper)
summary(model.hyper)

# Testing hypermutated samples only with log transformation
model.hyper.log10 = lm(log10(Non.silentMutationsperMb) ~ WGD + Type, data=mut.rates.hyper)
summary(model.hyper.log10)

# Testing non-hypermutated samples only
model.nonhyper = lm(Non.silentMutationsperMb ~ WGD + Type, data=mut.rates.nonhyper)
summary(model.nonhyper)

# Testing non-hypermutated samples only with log transformation
model.nonhyper.log10 = lm(log10(Non.silentMutationsperMb) ~ WGD + Type, data=mut.rates.nonhyper)
summary(model.nonhyper.log10)
```



Next, we calculate the differences between WGD and non-WGD samples within each tumor type using the non-parametric Wilcoxon rank-sum test.

```{r type_mut_burden_wilcox}
res = c()
tumor.type = unique(unique(mut.rates.nonhyper$Type))
for(i in tumor.type) {
  mut.rates.type = subset(mut.rates.nonhyper, Type == i)
  model = wilcox.test(Non.silentMutationsperMb ~ WGD, data = mut.rates.type)
  agg.median = aggregate(mut.rates.type$Non.silentMutationsperMb,
                  by=list(mut.rates.type$WGD),
                  median)
  agg.mean = aggregate(mut.rates.type$Non.silentMutationsperMb,
                  by=list(mut.rates.type$WGD),
                  mean)
  res = rbind(res, c(agg.median[,2], agg.mean[,2], model$p.value))
}
colnames(res) = c("Non-WGD_Median", "WGD_Median",
                  "Non-WGD_Mean", "WGD_Mean",
                  "Pvalue")
res = data.frame(Type = tumor.type,
                 res,
                 FDR=p.adjust(res[,"Pvalue"], "fdr"),
                 Bonferroni = p.adjust(res[,"Pvalue"], "bon"),
                 check.names = FALSE)

res = res[order(res$Pvalue),]

# Show results in table
kable(res, style = 'html', row.names = FALSE) %>%
  kable_styling(bootstrap_options = "striped") %>%
  scroll_box(width = "100%", height = "800px")
```



We can plot the mutation counts per Mb for each tumor type divided into WGD and non-WGD. 

```{r WGD_Violin_Global}
ggplot(mut.rates.nonhyper, aes(x=Type, y=Non.silentMutationsperMb, fill=WGD)) +
  geom_violin() + 
  theme_bw() +
  scale_y_log10() +
  ylab("Non-silent mutations per Mb (log10)") + 
  xlab("Tumor Type") + 
  theme(axis.text.x = element_text(angle = 45, hjust = 1),
        panel.grid.major.x = element_blank(),
        panel.grid.minor.x = element_blank())

```



We can also plot each tumor type individually to better see the differences between WGD and non-WGD tumors.

```{r WGD_Violin_Nonhyper}
ggplot(mut.rates.nonhyper, aes(x=WGD, y=Non.silentMutationsperMb, fill=WGD)) +
  geom_violin() + 
  geom_jitter(size = 0.11, color="grey") +
  facet_wrap(~ Type) + 
  theme_bw() +
  scale_y_log10() +
  ylab("Non-silent mutations per Mb (log10)") + 
  theme(axis.text.x = element_blank(),
        axis.ticks.x = element_blank(),
        panel.grid.major.x = element_blank(),
        panel.grid.minor.x = element_blank())
```

# Testing associations between mutation burden and WGD status while controlling for ploidy

We perform the same analysis as in the previous section, but add Ploidy as a covariate to correct for overall DNA content regardless of whether the increase in ploidy was due to WGD or not. 

```{r mut_rates_ploidy}
# Controlling for tumor type and Hypermutation
model.full.ploidy = lm(Non.silentMutationsperMb ~ WGD + Ploidy + MSI_POLE_Status + Type, data=mut.rates)
summary(model.full.ploidy)

# Controlling for tumor type and Hypermutation with log transformation
model.full.log10.ploidy = lm(log10(Non.silentMutationsperMb) ~ WGD + Ploidy + MSI_POLE_Status + Type, data=mut.rates)
summary(model.full.log10.ploidy)

# Testing hypermutated samples only
model.hyper.ploidy = lm(Non.silentMutationsperMb ~ WGD + Ploidy + Type, data=mut.rates.hyper)
summary(model.hyper.ploidy)

# Testing hypermutated samples only with log transformation
model.hyper.log10.ploidy = lm(log10(Non.silentMutationsperMb) ~ WGD + Ploidy + Type, data=mut.rates.hyper)
summary(model.hyper.log10.ploidy)

# Testing non-hypermutated samples only
model.nonhyper.ploidy = lm(Non.silentMutationsperMb ~ WGD + Ploidy + Type, data=mut.rates.nonhyper)
summary(model.nonhyper.ploidy)

# Testing non-hypermutated samples only with log transformation
model.nonhyper.log10.ploidy = lm(log10(Non.silentMutationsperMb) ~ WGD + Ploidy + Type, data=mut.rates.nonhyper)
summary(model.nonhyper.log10.ploidy)
```

In the last model examining non-hypermutated samples, we can see that mutation burden is still associated with WGD even after correcting for overall ploidy. Note that although the p-value is signficant the overall effect size is still relatively low, 0.07 in log10 scale is ~1.17, so WGD seems to increase mutation burden by about 1 per MB independent of DNA-content.  

Next, we calculate the differences between WGD and non-WGD samples within each tumor type using the non-parametric Wilcoxon rank-sum test. In contrast to previous section, we divided the log10 mutation burden by the tumor's ploidy to correction for overall DNA content. 

```{r type_mut_burden_wilcox_ploidy}
res.ploidy = c()
tumor.type = unique(unique(mut.rates.nonhyper$Type))
for(i in tumor.type) {
  mut.rates.type = subset(mut.rates.nonhyper, Type == i)
  model = wilcox.test(PloidyCorrected_MutationPerMb ~ WGD, data = mut.rates.type)
  agg.median = aggregate(mut.rates.type$PloidyCorrected_MutationPerMb,
                  by=list(mut.rates.type$WGD),
                  median)
  agg.mean = aggregate(mut.rates.type$PloidyCorrected_MutationPerMb,
                  by=list(mut.rates.type$WGD),
                  mean)
  res.ploidy = rbind(res.ploidy, c(agg.median[,2], agg.mean[,2], model$p.value))
}
colnames(res.ploidy) = c("Non-WGD_Median", "WGD_Median",
                  "Non-WGD_Mean", "WGD_Mean",
                  "Pvalue")
res.ploidy = data.frame(Type = tumor.type,
                 res.ploidy,
                 FDR=p.adjust(res.ploidy[,"Pvalue"], "fdr"),
                 Bonferroni = p.adjust(res.ploidy[,"Pvalue"], "bon"),
                 check.names = FALSE)

res.ploidy = res.ploidy[order(res.ploidy$Pvalue),]

# Show results in table
kable(res.ploidy, style = 'html', row.names = FALSE) %>%
  kable_styling(bootstrap_options = "striped") %>%
  scroll_box(width = "100%", height = "800px")
```



We can plot the mutation counts per Mb divided by ploidy for each tumor type divided into WGD and non-WGD. 


```{r WGD_Violin_Global_PloidyCorrected}
ggplot(mut.rates.nonhyper, aes(x=Type, y=PloidyCorrected_MutationPerMb, fill=WGD)) +
  geom_violin() + 
  theme_bw() +
  ylab("Non-silent mutations per Mb (log10) / Ploidy") + 
  xlab("Tumor Type") + 
  theme(axis.text.x = element_text(angle = 45, hjust = 1),
        panel.grid.major.x = element_blank(),
        panel.grid.minor.x = element_blank())

```

We can also plot each tumor type individually to better see the differences between Ploidy-corrected mutations burden and WGD/non-WGD tumors.


```{r WGD_Violin_Nonhyper_Ploidy}
ggplot(mut.rates.nonhyper, aes(x=WGD, y=PloidyCorrected_MutationPerMb, fill=WGD)) +
  geom_violin() + 
  geom_jitter(size = 0.11, color="grey") +
  facet_wrap(~ Type) + 
  theme_bw() +
  ylab("Non-silent mutations per Mb (log10) / Ploidy") + 
  theme(axis.text.x = element_blank(),
        axis.ticks.x = element_blank(),
        panel.grid.major.x = element_blank(),
        panel.grid.minor.x = element_blank())
```

# Associations between gene mutational frequencies and WGD status

## All tumors

```{r LR_all}
muts.nonhyper = muts[,rownames(mut.rates.nonhyper)]
muts.nonhyper.min10 = muts.nonhyper[rowSums(muts.nonhyper) > 10,]

res.gene.lr = logistic.regression(muts.nonhyper.min10, ~ WGD + Log10_MutsPerMB + Type, covar = mut.rates.nonhyper, verbose = FALSE)
res.gene.lr.summary = data.frame(Gene = rownames(res.gene.lr$Estimate),                                  WGD_Estimate=res.gene.lr$Estimate[,"WGDWGD"],
                          WGD_Zvalue = res.gene.lr$Zvalue[,"WGDWGD"],
                          WGD_Pvalue = res.gene.lr$Pvalue[,"WGDWGD"],
                          WGD_FDR = res.gene.lr$FDR[,"WGDWGD"])
res.gene.lr.summary = res.gene.lr.summary[order(res.gene.lr.summary$WGD_Pvalue),]
write.table(res.gene.lr.summary, "LR_WGD_Results_Nonhyper.txt", sep="\t", quote=FALSE, row.names=FALSE)

label = rownames(res.gene.lr.summary)
label[res.gene.lr.summary$WGD_FDR > 0.05] = ""

# Plot with all genes
g1 = ggplot(res.gene.lr.summary, aes(x=WGD_Estimate,
                                y=-log10(WGD_Pvalue),
                                color = WGD_FDR < 0.05)) +
  geom_point() +
  theme_bw() + xlab("Log odds ratio") + ylab("P-value (-log10)") +
  theme(panel.grid = element_blank()) + 
  geom_vline(xintercept = 0, lty=2, color="grey") +
  geom_text_repel(label = label) + 
  ggtitle("Plotting all genes")

# Plot excluding TP53
g2 = ggplot(res.gene.lr.summary[-1,], aes(x=WGD_Estimate,
                                y=-log10(WGD_Pvalue),
                                color = WGD_FDR < 0.05)) +
  geom_point() +
  theme_bw() + xlab("Log odds ratio") + ylab("P-value (-log10)") +
  theme(panel.grid = element_blank()) + 
  geom_vline(xintercept = 0, lty=2, color="grey") +
  geom_text_repel(label = label[-1]) + 
  ggtitle("Excluding TP53 from plot")

plot_grid(g1, g2, labels = "AUTO")

# Show table of results
res.gene.lr.summary.sig = subset(res.gene.lr.summary, WGD_FDR < 0.25)
kable(res.gene.lr.summary.sig, style = 'html', row.names = FALSE) %>%
  kable_styling(bootstrap_options = "striped") %>%
  scroll_box(width = "100%", height = "800px")
```

## Tumors without TP53 mutations

```{r LR_nonTP53}
tumor.non.tp53 = colnames(muts.nonhyper)[muts.nonhyper["TP53",] == 0]
i = intersect(intersect(rownames(mut.rates.nonhyper), colnames(muts.nonhyper)), tumor.non.tp53)
mut.rates.nonTP53 = mut.rates.nonhyper[i,]
muts.nonTP53 = muts.nonhyper[,i]

muts.nonTP53.min10 = muts.nonTP53[rowSums(muts.nonTP53) > 10,]

res.gene.lr.nonTP53 = logistic.regression(muts.nonTP53.min10, ~ WGD + Log10_MutsPerMB + Type, covar = mut.rates.nonTP53, verbose = FALSE)

res.gene.lr.nonTP53.summary = data.frame(Gene = rownames(res.gene.lr.nonTP53$Estimate),                                  WGD_Estimate=res.gene.lr.nonTP53$Estimate[,"WGDWGD"],
                          WGD_Zvalue = res.gene.lr.nonTP53$Zvalue[,"WGDWGD"],
                          WGD_Pvalue = res.gene.lr.nonTP53$Pvalue[,"WGDWGD"],
                          WGD_FDR = res.gene.lr.nonTP53$FDR[,"WGDWGD"])
res.gene.lr.nonTP53.summary = res.gene.lr.nonTP53.summary[order(res.gene.lr.nonTP53.summary$WGD_Pvalue),]
write.table(res.gene.lr.nonTP53.summary, "LR_WGD_Results_Nonhyper_nonTP53mutated.txt", sep="\t", quote=FALSE, row.names=FALSE)

label = rep("", nrow(res.gene.lr.nonTP53.summary))
top = head(order(res.gene.lr.summary$WGD_Pvalue), 10)
label[top] = rownames(res.gene.lr.nonTP53.summary)[top]

# Plot with all significant genes
g3 = ggplot(res.gene.lr.nonTP53.summary, aes(x=WGD_Estimate,
                                y=-log10(WGD_Pvalue),
                                color = WGD_FDR < 0.05)) +
  geom_point() +
  theme_bw() + xlab("Log odds ratio") + ylab("P-value (-log10)") +
  theme(panel.grid = element_blank()) + 
  geom_vline(xintercept = 0, lty=2, color="grey") +
  geom_text_repel(label = label) +
  xlim(-3, 3) +
  ggtitle("Including only non-TP53 mutated samples")
g3

# Show table of results
res.gene.lr.nonTP53.summary.sig = subset(res.gene.lr.nonTP53.summary, WGD_Pvalue < 0.05)
kable(res.gene.lr.nonTP53.summary.sig, style = 'html', row.names = FALSE) %>%
  kable_styling(bootstrap_options = "striped") %>%
  scroll_box(width = "100%", height = "800px")
```


## Tumors with TP53 mutations

```{r LR_withTP53}
tumor.tp53 = colnames(muts.nonhyper)[muts.nonhyper["TP53",] > 0]
i = intersect(intersect(rownames(mut.rates.nonhyper), colnames(muts.nonhyper)), tumor.tp53)

# Determine the number of tumor types that will have at least 5 tumors
# after subsetting to TP53 mutants
type.table = table(mut.rates.nonhyper[i,"Type"])
type.n = names(type.table)[type.table > 4]
tumor.type.n = rownames(subset(mut.rates.nonhyper, Type %in% type.n))

i2 = intersect(intersect(intersect(rownames(mut.rates.nonhyper), colnames(muts.nonhyper)), tumor.tp53), tumor.type.n)

mut.rates.TP53 = mut.rates.nonhyper[i2,]
mut.rates.TP53$Type = droplevels(mut.rates.TP53$Type)
muts.TP53 = muts.nonhyper[,i2]
muts.TP53.min10 = muts.TP53[rowSums(muts.TP53) > 10,]


res.gene.lr.TP53 = logistic.regression(muts.TP53.min10, ~ WGD + Log10_MutsPerMB + Type, covar = mut.rates.TP53, verbose = FALSE)

res.gene.lr.TP53.summary = data.frame(Gene = rownames(res.gene.lr.TP53$Estimate),                                  WGD_Estimate=res.gene.lr.TP53$Estimate[,"WGDWGD"],
                          WGD_Zvalue = res.gene.lr.TP53$Zvalue[,"WGDWGD"],
                          WGD_Pvalue = res.gene.lr.TP53$Pvalue[,"WGDWGD"],
                          WGD_FDR = res.gene.lr.TP53$FDR[,"WGDWGD"])
res.gene.lr.TP53.summary = res.gene.lr.TP53.summary[order(res.gene.lr.TP53.summary$WGD_Pvalue),]
write.table(res.gene.lr.TP53.summary, "LR_WGD_Results_Nonhyper_TP53mutated.txt", sep="\t", quote=FALSE, row.names=FALSE)

label = rep("", nrow(res.gene.lr.TP53.summary))
top = head(order(res.gene.lr.summary$WGD_Pvalue), 10)
label[top] = rownames(res.gene.lr.TP53.summary)[top]

# Plot with all significant genes
g4 = ggplot(res.gene.lr.TP53.summary, aes(x=WGD_Estimate,
                                y=-log10(WGD_Pvalue),
                                color = WGD_FDR < 0.05)) +
  geom_point() +
  theme_bw() + xlab("Log odds ratio") + ylab("P-value (-log10)") +
  theme(panel.grid = element_blank()) + 
  geom_vline(xintercept = 0, lty=2, color="grey") +
  geom_text_repel(label = label) +
  xlim(-5, 5) +
  ggtitle("Including only TP53 mutated samples")
g4

# Show table of results
res.gene.lr.nonTP53.summary.sig = subset(res.gene.lr.nonTP53.summary, WGD_Pvalue < 0.05)
kable(res.gene.lr.nonTP53.summary.sig, style = 'html', row.names = FALSE) %>%
  kable_styling(bootstrap_options = "striped") %>%
  scroll_box(width = "100%", height = "800px")
```

# Session Information
```{r session}
sessionInfo()
```

